File: src/jsMain/resources/VizPanel.js
class VizPanel {
  constructor(panel) {
    this.panel = panel;
    this.name = panel.name;
    this.geometry = new THREE.Geometry();
    const panelGeometry = this.geometry;
    const panelVertices = panelGeometry.vertices;

    let triangle = new THREE.Triangle(); // for computing area...
    this.area = 0;

    panelGeometry.faces = panel.faces.faces.toArray().map(face => {
      let localVerts = [];
      face.vertexIds.toArray().forEach(vi => {
        let v = geom.vertices[vi];
        let lvi = panelVertices.indexOf(v);
        if (lvi === -1) {
          lvi = panelVertices.length;
          panelVertices.push(v);
        }
        localVerts.push(lvi)
      });

      triangle.set(...localVerts.map(x => panelVertices[x]));
      this.area += triangle.getArea();

      return new THREE.Face3(...localVerts);
    });

    this.isMultiFaced = panelGeometry.faces.length > 1;

    this.edgeNeighbors = {};
    panelGeometry.faces.forEach(face => {
      [[face.a, face.b], [face.b, face.c], [face.c, face.a]].forEach(vs => {
        vs = vs.sort().join("-");
        let neighbors = this.edgeNeighbors[vs];
        if (neighbors === undefined) {
          neighbors = [];
          this.edgeNeighbors[vs] = neighbors;
        }
        neighbors.push(face);
      });
    });

    geom.computeVertexNormals(); // todo: why is this here?

    let lines = panel.lines.toArray().map(line => {
      let lineGeo = new THREE.Geometry();
      lineGeo.vertices = line.points.toArray().map(pt => new THREE.Vector3(pt.x, pt.y, pt.z));
      return lineGeo;
    });

    this.faceMaterial = new THREE.MeshBasicMaterial({color: 0xaa0000,});
    this.faceMaterial.side = THREE.FrontSide;
    this.faceMaterial.transparent = true;
    this.faceMaterial.opacity = 0.99;

    this.mesh = new THREE.Mesh(panelGeometry, this.faceMaterial);
    this.mesh.panel = this; // so we can get back to the VizPanel from a raycaster intersection...
    this.mesh.visible = false;

    this.lines = lines.map(line => new THREE.Line(line, lineMaterial));

    scene.add(this.mesh);
    this.lines.forEach((line) => {
      scene.add(line);
    });
  }

  addPixels(pixelCount) {
    const panelGeometry = this.geometry;
    const vertices = panelGeometry.vertices;
    panelGeometry.computeFaceNormals();
    const pixelsGeometry = new THREE.Geometry();

    let quaternion = new THREE.Quaternion();

    const panelFaces = panelGeometry.faces;
    let curFace = panelFaces[0];
    let revertToNormal = curFace.normal.clone();
    let straightOnNormal = new THREE.Vector3(0, 0, 1);
    quaternion.setFromUnitVectors(curFace.normal, straightOnNormal);
    let matrix = new THREE.Matrix4();
    matrix.makeRotationFromQuaternion(quaternion);
    panelGeometry.applyMatrix(matrix);
    pixelsGeometry.applyMatrix(matrix);

    let pixelSpacing = 2; // inches
    let pos = this.randomLocation(curFace, vertices);
    const nextPos = new THREE.Vector3();

    pixelsGeometry.vertices.push(pos.clone());
    const colors = [];
    colors.push(0, 0, 0);

    let tries = 1000;
    let angleRad = Math.random() * 2 * Math.PI;
    let angleRadDelta = Math.random() * 0.5 - 0.5;
    let pixelsSinceEdge = 0;
    for (let pixelI = 1; pixelI < pixelCount; pixelI++) {
      nextPos.x = pos.x + pixelSpacing * Math.sin(angleRad);
      nextPos.y = pos.y + pixelSpacing * Math.cos(angleRad);
      nextPos.z = pos.z;

      // console.log("cur face: ", this.faceVs(curFace, panelGeometry));

      if (!this.isInsideFace(curFace, nextPos)) {
        let newFace = this.getFaceForPoint(curFace, nextPos);
        if (newFace) {
          // console.log("moving from", curFace, "to", newFace);
          // console.log("prior face vs:", this.faceVs(curFace, panelGeometry));

          quaternion.setFromUnitVectors(straightOnNormal, revertToNormal);
          matrix.makeRotationFromQuaternion(quaternion);
          panelGeometry.applyMatrix(matrix);
          pixelsGeometry.applyMatrix(matrix);
          nextPos.applyMatrix4(matrix);

          curFace = newFace;
          revertToNormal = curFace.normal.clone();
          quaternion.setFromUnitVectors(curFace.normal, straightOnNormal);
          matrix.makeRotationFromQuaternion(quaternion);
          panelGeometry.applyMatrix(matrix);
          pixelsGeometry.applyMatrix(matrix);
          // console.log("pos was", nextPos);
          nextPos.applyMatrix4(matrix);
          // console.log("pos is now", nextPos);
          // console.log("new face vs:", this.faceVs(newFace, panelGeometry));
          nextPos.z = panelGeometry.vertices[newFace.a].z;
          if (!this.isInsideFace(curFace, nextPos)) {
            // console.log(nextPos, "is not in", this.faceVs(curFace, panelGeometry));
            nextPos.copy(this.randomLocation(curFace, vertices));
          } else {
            // console.log("AWESOME", nextPos, "is in", this.faceVs(curFace, panelGeometry))
          }
        } else {
          angleRad = Math.random() * 2 * Math.PI;
          pixelI--;
          if (tries-- < 0) break;
          pixelsSinceEdge = 0;
          continue;
        }
      }

      // console.log("pixel z = ", nextPos.z);
      pixelsGeometry.vertices.push(nextPos.clone());
      colors.push(0, 0, 0);

      angleRad += angleRadDelta;
      angleRadDelta *= 1 - Math.random() * 0.2 + 0.1;

      // occasional disruption just in case we're in a tight loop...
      if (pixelsSinceEdge > pixelCount / 10) {
        angleRad = Math.random() * 2 * Math.PI;
        angleRadDelta = Math.random() * 0.5 - 0.5;
        pixelsSinceEdge = 0;
      }
      pos.copy(nextPos);
      pixelsSinceEdge++;
    }

    quaternion.setFromUnitVectors(straightOnNormal, revertToNormal);
    matrix.makeRotationFromQuaternion(quaternion);
    panelGeometry.applyMatrix(matrix);
    pixelsGeometry.applyMatrix(matrix);

    const pixBufGeometry = new THREE.BufferGeometry();
    pixBufGeometry.addAttribute('position',
        new THREE.Float32BufferAttribute(pixelsGeometry.vertices.flatMap(v => [v.x, v.y, v.z]), 3));

    let colorsBuffer = new THREE.Float32BufferAttribute(colors, 3);
    colorsBuffer.dynamic = true;
    pixBufGeometry.addAttribute('color', colorsBuffer);
    const material = new THREE.PointsMaterial({size: 3, vertexColors: THREE.VertexColors});
    const points = new THREE.Points(pixBufGeometry, material);
    scene.add(points);

    this.pixelCount = pixelCount;
    this.pixelColorsBuffer = colorsBuffer;
    this.pixelsGeometry = pixBufGeometry;
  }

  faceVs(face, geometry) {
    return [face.a, face.b, face.c].map(f => geometry.vertices[f]);
  }

  faceZs(face, geometry) {
    return [face.a, face.b, face.c].map(f => geometry.vertices[f].z);
  }

  randomLocation(face, vertices) {
    const v = new THREE.Vector3().copy(vertices[face.a]);
    v.addScaledVector(new THREE.Vector3().copy(vertices[face.b]).sub(v), Math.random());
    v.addScaledVector(new THREE.Vector3().copy(vertices[face.c]).sub(v), Math.random());
    return v;
  }

  isInsideFace(curFace, v) {
    const vertices = this.geometry.vertices;

    return VizPanel.isInside(VizPanel.xy(v), [
      VizPanel.xy(vertices[curFace.a]),
      VizPanel.xy(vertices[curFace.b]),
      VizPanel.xy(vertices[curFace.c])]);
  }

  static isInside(point, vs) {
    // ray-casting algorithm based on
    // https://wrf.ecse.rpi.edu/Research/Short_Notes/pnpoly.html

    var x = point[0], y = point[1];

    var inside = false;
    for (var i = 0, j = vs.length - 1; i < vs.length; j = i++) {
      var xi = vs[i][0], yi = vs[i][1];
      var xj = vs[j][0], yj = vs[j][1];

      var intersect = ((yi > y) != (yj > y))
          && (x < (xj - xi) * (y - yi) / (yj - yi) + xi);
      if (intersect) {
        inside = !inside;
      }
    }

    return inside;
  }

  static xy(v) {
    return [v.x, v.y];
  }

  // we've tried to add a pixel that's not inside curFace; figure out which face it corresponds to...
  getFaceForPoint(curFace, v) {
    if (this.isMultiFaced) {
      const vertices = this.geometry.vertices;

      // find the edge closest to v...
      let closestEdge = [-1, -1];
      let bestDistance = Infinity;
      [[curFace.a, curFace.b], [curFace.b, curFace.c], [curFace.c, curFace.a],].forEach(edgeVs => {
        const closestPointOnEdge = new THREE.Vector3();
        const v0 = edgeVs[0];
        const v1 = edgeVs[1];
        new THREE.Line3(vertices[v0], vertices[v1]).closestPointToPoint(v, true, closestPointOnEdge);
        let thisDistance = closestPointOnEdge.distanceTo(v);
        if (thisDistance < bestDistance) {
          closestEdge = [v0, v1];
          bestDistance = thisDistance;
        }
      });

      let edgeId = closestEdge.sort().join("-");
      // console.log("Closest edge to", v, "is", edgeId, this.edgeNeighbors[edgeId]);

      const neighbors = this.edgeNeighbors[edgeId];
      const neighbor = neighbors.filter(f => f !== curFace);
      if (neighbor.length === 0) {
        return null;
      } else if (neighbor.length > 1) {
        // console.warn("Found multiple neighbors for ", this.panel.name, " edge ", edgeId, ": ", neighbors);
      }

      // console.log("Face for ", v, "is", edgeId, neighbor[0]);
      return neighbor[0];
    }
    return null;
  }

  setPanelColor(panelBgColor, pixelColors) {
    this.mesh.visible = true;

    if (!renderPixels) {
      this.faceMaterial.color.r = panelBgColor.redF;
      this.faceMaterial.color.g = panelBgColor.greenF;
      this.faceMaterial.color.b = panelBgColor.blueF;
    } else {
      this.faceMaterial.color.r = .3;
      this.faceMaterial.color.g = .3;
      this.faceMaterial.color.b = .3;
    }

    if (this.pixelCount && pixelColors) {
      const count = Math.min(this.pixelCount, pixelColors.length);
      for (let i = 0; i < count; i++) {
        const pColor = pixelColors[i];
        this.pixelColorsBuffer.array[i * 3] = pColor.redF;
        this.pixelColorsBuffer.array[i * 3 + 1] = pColor.greenF;
        this.pixelColorsBuffer.array[i * 3 + 2] = pColor.blueF;
      }
      this.pixelColorsBuffer.needsUpdate = true;
    }
  }

}

File: src/jsMain/kotlin/baaahs/visualizer/VizPanel.kt
package baaahs.visualizer

import baaahs.Color
import baaahs.Model
import baaahs.Pixels
import baaahs.geom.Vector2
import info.laht.threekt.THREE.AdditiveBlending
import info.laht.threekt.THREE.FrontSide
import info.laht.threekt.THREE.VertexColors
import info.laht.threekt.core.BufferGeometry
import info.laht.threekt.core.Face3
import info.laht.threekt.core.Geometry
import info.laht.threekt.core.Object3D
import info.laht.threekt.geometries.PlaneBufferGeometry
import info.laht.threekt.loaders.TextureLoader
import info.laht.threekt.materials.LineBasicMaterial
import info.laht.threekt.materials.MeshBasicMaterial
import info.laht.threekt.math.Triangle
import info.laht.threekt.math.Vector3
import info.laht.threekt.math.minus
import info.laht.threekt.objects.Line
import info.laht.threekt.objects.Mesh
import info.laht.threekt.scenes.Scene
import org.khronos.webgl.Float32Array
import org.khronos.webgl.get
import org.khronos.webgl.set
import org.w3c.dom.get
import three.BufferGeometryUtils
import three.Float32BufferAttribute
import three.Matrix4
import kotlin.browser.document
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

class VizPanel(panel: Model.Surface, private val geom: Geometry, private val scene: Scene) {
    companion object {
        private val roundLightTx = TextureLoader().load(
            "./visualizer/textures/round.png",
            { println("loaded!") },
            { println("progress!") },
            { println("error!") }
        )

        fun getFromObject(object3D: Object3D): VizPanel? =
            object3D.userData.asDynamic()["VizPanel"] as VizPanel?
    }

    val name = panel.name
    internal val geometry = Geometry()
    var area = 0.0f
    private var panelNormal: Vector3
    val isMultiFaced: Boolean
    internal var edgeNeighbors: Map<String, List<Face3>>
    private val lineMaterial = LineBasicMaterial().apply { color.set(0xaaaaaa) }
    internal var faceMaterial: MeshBasicMaterial
    private var mesh: Mesh
    private var lines: List<Line>
    var vizPixels: VizPixels? = null
        set(value) {
            field?.removeFromScene(scene)
            value?.addToScene(scene)

            field = value
        }

    init {
        val panelGeometry = this.geometry
        val panelVertices = panelGeometry.vertices

        val triangle = Triangle() // for computing area...

        val faceAreas = mutableListOf<Float>()
        panelGeometry.faces = panel.faces.map { face ->
            val localVerts = face.vertexIds.map { vi ->
                val v = geom.vertices[vi]
                var lvi = panelVertices.indexOf(v)
                if (lvi == -1) {
                    lvi = panelVertices.size
                    panelVertices.asDynamic().push(v)
                }
                lvi
            }

            triangle.set(
                panelVertices[localVerts[0]],
                panelVertices[localVerts[1]],
                panelVertices[localVerts[2]]
            )

            val faceArea = triangle.asDynamic().getArea() as Float
            faceAreas.add(faceArea)
            this.area += faceArea

            val normal: Vector3 = document["non-existant-key"]
            Face3(localVerts[0], localVerts[1], localVerts[2], normal)
        }.toTypedArray()

        isMultiFaced = panelGeometry.faces.size > 1

        panelGeometry.computeFaceNormals()
        val faceNormalSum = Vector3()
        panelGeometry.faces.forEachIndexed { index, face ->
            val faceArea = faceAreas[index]
            faceNormalSum.addScaledVector(face.normal!!, faceArea)
        }
        panelNormal = faceNormalSum.divideScalar(area.toDouble())

        val edgeNeighbors = mutableMapOf<String, MutableList<Face3>>()
        panelGeometry.faces.forEach { face ->
            face.segments().forEach { vs ->
                val vsKey = vs.asKey()
                val neighbors = edgeNeighbors.getOrPut(vsKey) { mutableListOf() }
                neighbors.add(face)
            }
        }
        this.edgeNeighbors = edgeNeighbors

        geom.computeVertexNormals() // todo: why is this here?

        val lines = panel.lines.map { line ->
            val lineGeo = Geometry()
            lineGeo.vertices = line.vertices.map { pt -> Vector3(pt.x, pt.y, pt.z) }.toTypedArray()
            lineGeo
        }

        this.faceMaterial = MeshBasicMaterial().apply { color.set(0x222222) }
        this.faceMaterial.side = FrontSide
        this.faceMaterial.transparent = false

        this.mesh = Mesh(panelGeometry, this.faceMaterial)
        mesh.asDynamic().name = "Surface: $name"

        // so we can get back to the VizPanel from a raycaster intersection:
        this.mesh.userData.asDynamic()["VizPanel"] = this

        scene.add(this.mesh)

        this.lines = lines.map { line -> Line(line.asDynamic(), lineMaterial) }

        this.lines.forEach { line ->
            scene.add(line)
        }
    }

    class Point2(val x: Float, val y: Float) {
        operator fun component1() = x
        operator fun component2() = y
    }


    class VizPixels(vizPanel: VizPanel, val positions: Array<Vector3>) : Pixels {
        override val size = positions.size
        private val pixGeometry = BufferGeometry()
        private val planeGeometry: BufferGeometry
        private val vertexColorBufferAttr: Float32BufferAttribute
        private val colorsAsInts = IntArray(size) // store colors as an int array too for Pixels.get()

        init {
            val positionsArray = Float32Array(size * 3)
            positions.forEachIndexed { i, v ->
                positionsArray[i * 3] = v.x.toFloat()
                positionsArray[i * 3 + 1] = v.y.toFloat()
                positionsArray[i * 3 + 2] = v.z.toFloat()
            }

            val positionsBufferAttr = Float32BufferAttribute(positionsArray, 3)
            pixGeometry.addAttribute("position", positionsBufferAttr)

            vertexColorBufferAttr = Float32BufferAttribute(Float32Array(size * 3 * 4), 3)
            vertexColorBufferAttr.dynamic = true

            val rotator = Rotator(Vector3(0, 0, 1), vizPanel.panelNormal)
            planeGeometry = BufferGeometryUtils.mergeBufferGeometries(positions.map { position ->
                val geometry = PlaneBufferGeometry(2 + Random.nextFloat() * 8, 2 + Random.nextFloat() * 8)
                rotator.rotate(geometry)
                geometry.translate(position.x, position.y, position.z)
                geometry
            }.toTypedArray())
            planeGeometry.addAttribute("color", vertexColorBufferAttr)
        }

        private val pixelsMesh = Mesh(planeGeometry, MeshBasicMaterial().apply {
            side = FrontSide
            transparent = true
            blending = AdditiveBlending
//            depthFunc = AlwaysDepth
            depthTest = false
            depthWrite = false
            vertexColors = VertexColors

            map = roundLightTx
        })

        fun addToScene(scene: Scene) {
            scene.add(pixelsMesh)
        }

        fun removeFromScene(scene: Scene) {
            scene.remove(pixelsMesh)
        }

        override fun get(i: Int): Color {
            return Color(colorsAsInts[i])
        }

        override fun set(i: Int, color: Color) {
            colorsAsInts[i] = color.argb

            val redF = color.redF / 2
            val greenF = color.greenF / 2
            val blueF = color.blueF / 2

            val rgb3Buf = vertexColorBufferAttr
            rgb3Buf.setXYZ(i * 4, redF, greenF, blueF)
            rgb3Buf.setXYZ(i * 4 + 1, redF, greenF, blueF)
            rgb3Buf.setXYZ(i * 4 + 2, redF, greenF, blueF)
            rgb3Buf.setXYZ(i * 4 + 3, redF, greenF, blueF)
            vertexColorBufferAttr.needsUpdate = true
        }

        override fun set(colors: Array<Color>) {
            val maxCount = min(this.size, colors.size)
            val rgbBuf = vertexColorBufferAttr.array
            for (i in 0 until maxCount) {
                colorsAsInts[i] = colors[i].argb

                val pColor = colors[i]
                rgbBuf[i * 3] = pColor.redF / 2
                rgbBuf[i * 3 + 1] = pColor.greenF / 2
                rgbBuf[i * 3 + 2] = pColor.blueF / 2
            }
            vertexColorBufferAttr.needsUpdate = true
        }

        fun getPixelLocationsInModelSpace(vizPanel: VizPanel): Array<Vector3> = positions

        fun getPixelLocationsInPanelSpace(vizPanel: VizPanel): Array<Vector2> {
            val panelGeom = vizPanel.geometry.clone()
            val pixGeom = pixGeometry.clone()

            val straightOnNormal = Vector3(0, 0, 1)

            // Rotate to straight on.
            val rotator = Rotator(vizPanel.panelNormal, straightOnNormal)
            rotator.rotate(panelGeom)
            rotator.rotate(pixGeom)

            // Translate and scale pixels to panel space (0f..1f)
            panelGeom.computeBoundingBox()
            val boundingBox = panelGeom.boundingBox!!
            val min = boundingBox.min
            val size = boundingBox.max - boundingBox.min

            val translate = Matrix4().makeTranslation(-min.x, -min.y, -min.z)
            panelGeom.applyMatrix(translate)
            pixGeom.applyMatrix(translate)

            val scale = Matrix4().makeScale(1.0 / size.x, 1.0 / size.y, 1.0)
            panelGeom.applyMatrix(scale)
            pixGeom.applyMatrix(scale)

            val pixelVs = mutableListOf<Vector2>()
            val pixelPositions = pixGeom.getAttribute("position")
            val array = pixelPositions.array as Float32Array
            for (i in 0 until pixelPositions.count * 3 step 3) {
                val v = Vector2(clamp(array[i]).toDouble(), clamp(array[i + 1]).toDouble())
                pixelVs.add(v)
            }

            return pixelVs.toTypedArray()
        }

        fun clamp(f: Float): Float = min(1f, max(f, 0f))
    }

    fun getPixelLocationsInPanelSpace(): Array<Vector2>? {
        return vizPixels?.getPixelLocationsInPanelSpace(this)
    }

    fun getPixelLocationsInModelSpace(): Array<Vector3>? {
        return vizPixels?.getPixelLocationsInModelSpace(this)
    }
}
